### PR TITLE
feat: adds a custom react explorer

### PR DESCRIPTION
**Pull request by Encord-davids**

### CODE REVIEW

File: frontend_components/src/encord_active_components/frontend/src/components/explorer/Pagination.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  pageSize: number;
+  totalItems: number;
+  onChange: (to: number) => void;
+  onChangePageSize: (size: PageSize) => void;
+}) => {
+  const prev = current - 1;
+  const next = current + 1;
+
+  let totalPages = Math.max((totalItems / pageSize) | 0, 1);
+  if (totalItems % pageSize == 0) totalPages--;
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply frontend_components/src/encord_active_components/frontend/src/components/explorer/Pagination.tsx:38]: good catch!

File: frontend_components/src/encord_active_components/frontend/src/components/explorer/Pagination.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+const defaultPageSize = pageSizes[0];
+export type PageSize = typeof pageSizes[number];
+
+export const usePagination = <T extends any[]>(items: T) => {
+  const [page, setPage] = useState(1);
+  const [pageSize, setPageSize] = useState<PageSize>(defaultPageSize);
+
+  const pageItems = items.slice(
+    (page - 1) * pageSize,
+    (page + 1) * pageSize + pageSize
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply frontend_components/src/encord_active_components/frontend/src/components/explorer/Pagination.tsx:15]: yep absolutely

File: src/encord_active/server/main.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    )
+    app.mount("/static", StaticFiles(directory=path), name="static")
+    run(app)
+
+
+async def project(project: str):
+    return ProjectFileStructure(Path(environ.get("SERVER_START_PATH", "")) / project)
+
+
+Project = Annotated[ProjectFileStructure, Depends(project)]
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/server/main.py:55]: makes sense, what do you think about a `D` prefix for every dependency we inject? `DProjectFileStructure, DProject`

File: src/encord_active/lib/db/merged_metrics.py [hunk 4]
```diff
@@ -113,13 +113,18 @@ def update_tags(self, id: str, tags: List[Tag]):
...
-        return self._unsafe_all()
+    def all(self, marshall: bool = True, columns: Optional[List[str]] = None):
+        return self._unsafe_all(marshall, columns)
 
-    def _unsafe_all(self):
+    def _unsafe_all(self, marshall: bool, columns: Optional[List[str]]):
         with DBConnection() as conn:
-            merged_metrics = pd.read_sql(f"SELECT * FROM {TABLE_NAME}", conn, index_col="identifier")
-            merged_metrics.tags = merged_metrics.tags.apply(unmarshall_tags)
+            if columns and "identifier" not in columns:
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/lib/db/merged_metrics.py:121]: the opposite, if people select specific `columns` they have to also select the `identifier`

File: src/encord_active/lib/db/helpers/tags.py [hunk 5]
```diff
@@ -1,12 +1,14 @@
-from typing import Dict
+from typing import Dict, List
 
 import pandas as pd
+from encord_active_components.components.explorer import GroupedTags
 
-from encord_active.lib.db.tags import Tags
+from encord_active.app.common.components.tags.utils import all_tags
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/lib/db/helpers/tags.py:6]: yes yes

File: src/encord_active/lib/db/helpers/tags.py [hunk 6]
```diff
@@ -1,12 +1,14 @@
-from typing import Dict
+from typing import Dict, List
 
 import pandas as pd
+from encord_active_components.components.explorer import GroupedTags
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/lib/db/helpers/tags.py:4]: that is correct

File: src/encord_active/app/views/metrics.py [hunk 7]
```diff
@@ -40,11 +42,13 @@ def render():
...
                 return
 
+        start = perf_counter()
         selected_df = page.sidebar_options(available_metrics, metric_type)
 
         if selected_df is None:
             return
 
         page.build(selected_df, metric_type)
+        print("Done", perf_counter() - start)
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/app/views/metrics.py:52]: 👍

File: src/encord_active/app/data_quality/sub_pages/explorer.py [hunk 8]
```diff
@@ -85,20 +58,21 @@ def build(self, selected_df: DataFrame[MetricSchema], metric_scope: MetricScope)
...
             render_annotator_properties(selected_df)
 
-        fill_data_quality_window(selected_df, metric_scope, selected_metric)
+        output_state = UseState[Optional[Output]](None)
+        output = explorer(
+            project_name=get_state().project_paths.project_dir.name,
+            items=[id for id in get_state().filtering_state.merged_metrics.index.values],
+            scope=metric_scope.value,
+        )
+        if output and output != output_state:
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/app/data_quality/sub_pages/explorer.py:72]: this should just be removed

File: src/encord_active/app/common/state.py [hunk 9]
```diff
@@ -88,7 +87,7 @@ class State:
     reduced_embeddings: dict[EmbeddingType, Optional[DataFrame[Embedding2DSchema]]] = field(default_factory=dict)
 
     @classmethod
-    def init(cls, project_dir: Path, refresh_projects: Callable[[], Any]):
+    def init(cls, target_paht: Path, project_dir: Path, refresh_projects: Callable[[], Any]):
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/app/common/state.py:90]: 🙈

File: src/encord_active/app/common/state.py [hunk 10]
```diff
@@ -97,10 +96,10 @@ def init(cls, project_dir: Path, refresh_projects: Callable[[], Any]):
             PrismaConnection.set_project_path(project_dir)
             merged_metrics = MergedMetrics().all()
             st.session_state[StateKey.GLOBAL] = State(
+                target_path=target_paht,
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/app/common/state.py:99]: 🙈

File: src/encord_active/app/common/components/tags/utils.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+from encord_active.lib.db.tags import Tag, TagScope
+
+
+def scoped_tags(scopes: Set[TagScope]) -> List[Tag]:
+    return [tag for tag in all_tags() if tag.scope in scopes]
+
+
+def all_tags():
+    tag_list = MergedMetrics().all(columns=["tags"])["tags"].values
+    return list({tag for tag_list in tag_list for tag in tag_list})
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/app/common/components/tags/utils.py:13]: yas

File: src/encord_active/app/streamlit_entrypoint.py [hunk 12]
```diff
@@ -53,7 +53,7 @@ def provide_backcompatibility_for_old_predictions():
 
 def main(target: str):
     set_page_config()
-    target_path = Path(target)
+    target_path = Path(target).resolve()
```
[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/app/streamlit_entrypoint.py:56]: I think `expanduser + resolve` is the best

File: src/encord_active/server/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (frederik-encord) [reply src/encord_active/server/main.py:58]: Works for me. I used `Dep` in the premium api but `D` is fine as well.

[the remark this answers is not in the document]
Comment (Encord-davids) [reply src/encord_active/server/main.py:58]: just check the docs and seems that they are also postfixing with `Dep` so lets go with that

**Review by frederik-encord — CHANGES_REQUESTED** [review]
Not enough space: Perhaps shrink some buttons and add tooltips
image

Tooltips would be nice
image

When hovering the disabled button, it would be good to tell that one needs to select images first
image

When everything is selected, the inverse operation would be nice (deselect all)
image

Could we sort there? It's hard to find a specific metric in the list.
image

**Review by frederik-encord — DISMISSED** [review]

**Review by frederik-encord — DISMISSED** [review]

**Review by frederik-encord — APPROVED** [review]

**Review by frederik-encord — DISMISSED** [review]

**Review by frederik-encord — APPROVED** [review]

**Pull request merged**